RFC: a couple of `^` type fixes

nolta opened the pull request:
See the commit messages.

nolta committed:
don't promote Integer^BigInt to BigInt

Fixes the following inconsistency:

    julia> [0x2^T(9) for T in (Int16,Int32,Int64,BigInt)]
    4-element Array{Any,1}:
     0x00
     0x00
     0x00
      512

simonster commented on base/broadcast.jl:
```diff
@@ -316,8 +316,7 @@ function .//(A::AbstractArray, B::AbstractArray)
 
-type_pow(T,S) = promote_type(T,S)
-type_pow{S<:Integer}(::Type{Bool},::Type{S}) = Bool
+type_pow(T,S) = promote_type(Base.return_types(^,(T,S))...)
```
Won't this break type inference? Perhaps a staged function would be better?

JeffBezanson replied:
Something like `one(T)^one(S)` is always the best option if workable.

nolta replied:
Here's the reason i didn't use `typeof(one(T)^one(S))`:

```
julia> Real[2.1].^[2]
ERROR: InexactError()
 in setindex! at array.jl:328
 in _F_ at broadcast.jl:96
 in broadcast! at broadcast.jl:227
 in .^ at broadcast.jl:326
```

because `one(Real) === Int`.

nolta committed:
make ^ and .^ types consistent (part of #3274)

For example, before this patch:

    julia> typeof(int32(2) ^ int64(3))
    Int32

    julia> eltype([int32(2)] .^ [int64(3)])
    Int64  # Int32 after this patch

    julia> typeof(complex(2) ^ complex(3))
    Complex{Float64}

    julia> eltype([complex(2)] .^ [complex(3)])
    Complex{Int}  # Complex{Float64} after this patch

nolta commented on base/broadcast.jl:
```diff
@@ -316,8 +316,9 @@ function .//(A::AbstractArray, B::AbstractArray)
-type_pow{S<:Integer}(::Type{Bool},::Type{S}) = Bool
+stagedfunction type_pow(T,S)
+    :(promote_type(Base.return_types(^,(T,S))...))
+end
```
@simonster, is this what you had in mind?

joehuchette replied:
This is a really neat trick.

simonster replied:
Yes, this seems right to me.

nolta commented:
Rebased.
